internal/fs/fs.go

```diff
@@ -121,38 +178,160 @@ func genTestFilename(str string) string {
 	}, str)
 }
 
-// isCaseSensitiveFilesystem determines if the filesystem where dir
-// exists is case sensitive or not.
-//
-// CAVEAT: this function works by taking the last component of the given
-// path and flipping the case of the first letter for which case
-// flipping is a reversible operation (/foo/Bar → /foo/bar), then
-// testing for the existence of the new filename. There are two
-// possibilities:
-//
-// 1. The alternate filename does not exist. We can conclude that the
-// filesystem is case sensitive.
-//
-// 2. The filename happens to exist. We have to test if the two files
-// are the same file (case insensitive file system) or different ones
-// (case sensitive filesystem).
-//
-// If the input directory is such that the last component is composed
-// exclusively of case-less codepoints (e.g.  numbers), this function will
-// return false.
-func isCaseSensitiveFilesystem(dir string) bool {
-	alt := filepath.Join(filepath.Dir(dir),
-		genTestFilename(filepath.Base(dir)))
+var (
+	errSrcNotDir = errors.New("source is not a directory")
+	errDstExist  = errors.New("destination already exists")
+)
 
-	dInfo, err := os.Stat(dir)
+// CopyDir recursively copies a directory tree, attempting to preserve permissions.
+// Source directory must exist, destination directory must *not* exist.
+func CopyDir(src, dst string) error {
+	src = filepath.Clean(src)
+	dst = filepath.Clean(dst)
+
+	fi, err := os.Stat(src)
```

Comment:
ugh ugh symlinks.

Despite what was said earlier, I'm reconsidering this now - it seems unwise to just blithely include symlinks in this way. Especially because, as-written, this actually transforms the symlink into the referent, and continues the walk. I suspect that this approach could be an attack vector; it might run forever on a symlink loop, or happily walk back to / and starting trying to copy everything from there (which would then ultimately loop).

I think we need to change to relying on Lstat(), and copying the link itself, rather than transforming it into the referent.